### PR TITLE
Validate test runner

### PR DESCRIPTION
**Pull request by t-persson**
### Applicable Issues

fixes: eiffel-community/etos#156
This change derives from #38 and that commit can be ignored in this PR.

### Description of the Change

Verify that we can receive a digest via the docker repository APIs.

### Alternate Designs

I wanted to use the docker cli tool to do this but was not able to get it to work without doing `docker pull`, which we definitely don't want to do.

### Benefits

Tests will fail faster when the test runner does not exist leaving the users with less confusion.

### Possible Drawbacks

HTTP requests are, relatively, costly and something we have to do on every single run of ETOS. We might have to do some caching of the data if this becomes to heavy.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by: Tobias Persson <tobias.persson@axis.com>

### CODE REVIEW

**Review by andjoe-axis — CHANGES_REQUESTED** [review]

**Review by andjoe-axis — APPROVED** [review]

File: src/etos_api/library/docker.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    https://github.com/google/go-containerregistry/tree/main/cmd/crane
+    """
+
+    logger = logging.getLogger(__name__)
+    # In-memory database for stored authorization tokens.
+    # This dictionary shares memory with all instances of `Docker`, by design.
+    tokens = {}
+    lock = Lock()
+
+    def token(self, manifest_url):
```
Comment (fredjn) [inline src/etos_api/library/docker.py:41]: typehint?

File: src/etos_api/library/docker.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    def tag(self, base: str) -> tuple[str, str]:
+        """Figure out tag from a container image name.
+
+        :param base: Name of image.
+        :return: Base image name without tag and tag name.
+        """
+        tag = ""
+
+        parts = base.split(TAG_DELIMITER)
+        if len(parts) > 1 and REPO_DELIMITER not in parts[-1]:
```
Comment (fredjn) [inline src/etos_api/library/docker.py:108]: Add explanatory comment on code. E.g all the ins ans outs of ports and missing tags etc.

File: src/etos_api/library/docker.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+                "Probably found a registry URL in image name: (registry=%r, repo=%r)",
+                registry,
+                repo,
+            )
+        if registry in ("", "docker.io"):
+            self.logger.info("Assuming registry is %r for %r", DEFAULT_REGISTRY, repo)
+            registry = DEFAULT_REGISTRY
+        return registry, repo
+
+    async def digest(self, name):
```
Comment (fredjn) [inline src/etos_api/library/docker.py:141]: type hinting

**Review by fredjn — CHANGES_REQUESTED** [review]

File: src/etos_api/library/docker.py (hunk 2 above)
Comment (t-persson) [reply src/etos_api/library/docker.py:114]: Please check the comments added.

**Pull request merged**